src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,127 @@
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private Object[] currentList;
+    private Object[] newList;
+
+    public ArrayList() {
+        size = 0;
+        currentList = new Object[DEFAULT_CAPACITY];
+        newList = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == currentList.length) {
+            grow();
+            System.arraycopy(currentList, 0, newList, 0, size);
+            currentList = newList;
+        }
+        currentList[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (size == currentList.length) {
+            grow();
+        }
 
+        newList = new Object[newList.length];
+        System.arraycopy(currentList, 0, newList, 0, index);
+        newList[index] = value;
+        System.arraycopy(currentList, index, newList, index + 1, size - index);
+        size++;
+        currentList = newList;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (currentList.length < size + list.size()) {
+            newList = new Object[size + list.size()];
+            System.arraycopy(currentList, 0, newList, 0, size);
+            currentList = newList;
+        }
 
+        for (int i = 0; i < list.size(); i++) {
+            currentList[size++] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexChecking(index);
+        return (T) currentList[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexChecking(index);
+        currentList[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexChecking(index);
+
+        newList = new Object[newList.length];
+        System.arraycopy(currentList, 0, newList, 0, index);
+        System.arraycopy(currentList, index + 1, newList, index, size - index - 1);
+        newList[size - 1] = null;
+        size--;
+        T elementByIndex = (T) currentList[index];
+        currentList = newList;
+
+        return elementByIndex;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (element != null && element.equals(currentList[i])) {
+                index = i;
+                break;
+            }
+            if (element == null && currentList[i] == null) {
+                index = i;
+                break;
+            }
+        }
+
+        if (index == -1) {
+            throw new NoSuchElementException("There is no such element in the list");
+        }
```

Comment:
here it is better to loop through the old values, if you find it, you get the index and remove it through the remove (index) method